client/client.go

```diff
@@ -376,6 +367,67 @@ func (c *client) Close() {
 	}
 }
 
+func (c *client) setServiceMode(newMode pdpb.ServiceMode) {
+	c.svcModeMutex.Lock()
+	defer c.svcModeMutex.Unlock()
+
+	if newMode == c.serviceMode {
+		return
+	}
+
+	log.Info("changing service mode", zap.String("old-mode", pdpb.ServiceMode_name[int32(c.serviceMode)]),
+		zap.String("new-mode", pdpb.ServiceMode_name[int32(newMode)]))
+
+	if newMode == pdpb.ServiceMode_UNKNOWN_SVC_MODE {
+		log.Warn("intend to switch to unknown service mode. do nothing")
+		return
+	}
+
+	var newTSOCli *tsoClient
+	tsoSvcDiscovery := c.tsoSvcDiscovery
+	ctx, cancel := context.WithCancel(c.ctx)
+
+	if newMode == pdpb.ServiceMode_PD_SVC_MODE {
+		newTSOCli = newTSOClient(ctx, cancel, c.option, c.keyspaceID,
+			c.pdSvcDiscovery, c.pdSvcDiscovery.(tsoAllocatorEventSource), &pdTSOStreamBuilderFactory{})
+		newTSOCli.Setup()
+	} else {
+		tsoSvcDiscovery = newTSOServiceDiscovery(ctx, cancel, MetaStorageClient(c),
+			c.GetClusterID(c.ctx), c.keyspaceID, c.svrUrls, c.tlsCfg, c.option)
+		newTSOCli = newTSOClient(ctx, cancel, c.option, c.keyspaceID,
+			tsoSvcDiscovery, tsoSvcDiscovery.(tsoAllocatorEventSource), &tsoTSOStreamBuilderFactory{})
+		if err := tsoSvcDiscovery.Init(); err != nil {
+			cancel()
+			log.Error("failed to initialize tso service discovery. keep the current service mode",
+				zap.Strings("svr-urls", c.svrUrls), zap.String("current-mode", pdpb.ServiceMode_name[int32(c.serviceMode)]), zap.Error(err))
+			return
+		}
+		newTSOCli.Setup()
+	}
+
+	// cleanup the old tso client
+	if oldTSOCli := c.getTSOClient(); oldTSOCli != nil {
+		oldTSOCli.Close()
+	}
+	if c.serviceMode == pdpb.ServiceMode_API_SVC_MODE {
+		c.tsoSvcDiscovery.Close()
+	}
+
+	c.tsoSvcDiscovery = tsoSvcDiscovery
+	c.tsoClient.Store(newTSOCli)
+
+	log.Info("service mode changed", zap.String("old-mode", pdpb.ServiceMode_name[int32(c.serviceMode)]),
+		zap.String("new-mode", pdpb.ServiceMode_name[int32(newMode)]))
+	c.serviceMode = newMode
+}
+
+func (c *client) getTSOClient() *tsoClient {
+	if tsoCli := c.tsoClient.Load(); tsoCli != nil {
+		return tsoCli.(*tsoClient)
+	}
+	return nil
```

Comment:
Just returning `nil` is not a good idea as we have already met some similar panics before.

Reply:
This is to retrieve a pointer from an atomic value, so there is no need to differentiate the default value and non-existed cases. The caller anyway needs to check if the returned pointer is nil or not.

There are three places calling this function getTSOClient -- two places have checked if the returned pointer is nil and the third place doesn't check (I just added the nil check at the third place anyway). The third place doesn't check because once PD client is initialized, c.tsoClient will always have a non-nil pointer.